service/history/shard/context.go

```diff
@@ -73,9 +73,7 @@ type (
 		GenerateTaskID() (int64, error)
 		GenerateTaskIDs(number int) ([]int64, error)
 
-		GetTransferMaxReadLevel() int64
-		GetTimerMaxReadLevel(cluster string) time.Time
-		UpdateTimerMaxReadLevel(cluster string) time.Time
+		GetQueueMaxReadLevel(category persistence.HistoryTaskCategory, cluster string) persistence.HistoryTaskKey
```

Comment:
A get method doing updates is confusing. Let's rename this to make the intent explicit. `UpdateIfNeededAndGetQueueMaxReadLevel` or something like that. At least add comments about this behavior difference of timer vs transfer queue max levels